proto/pulumi/resource.proto

```diff
@@ -301,3 +324,7 @@ message Parameterization {
     string version = 2; // the parameterized package version.
     bytes value = 3; // the parameter value for the parameterized package.
 }
+
+message RegisterLifecycleHookRequest {
+    Callback callback = 1; // the callback that the engine can call to run the hook.
```

Comment:
Worth adding on_dry_run now? It's simple to support and almost certainly wanted.

Reply:
Happy to add it, but I thought we had discussed this in ideation and decided we didn't  need `on_dry_run` as an option, since you can call `pulumi.runtime.is_dry_run()` in the hook to determine if it should or shouldn't do something.

Reply:
I think by default hooks should not run at preview, I suspect most people won't deal with it correctly by default. So we want it opt-in and then once opted in you can see if it's a preview or not by looking at is_dry_run.

Reply:
Right, that makes sense. I've added the option in https://github.com/pulumi/pulumi/pull/19746/commits/ea1323321032800657fb25ee5b40aaba1747b6d8